File: src/build_env.cc
#include <eval.h>
#include <unistd.h>
#include <iostream>

static Value c_println(std::vector<Value> args) {
    for (const auto& arg : args) {
        std::cout << arg;
    }
    std::cout << "\n";
    return 0;
}

static Value c_print(std::vector<Value> args) {
    for (const auto& arg : args) {
        std::cout << arg;
    }
    return 0;
}

static Value c_input(std::vector<Value> args) {
    std::cout << args[0];
    std::string buf;
    std::getline(std::cin, buf);
    return buf;
}

static Value c_syscall(std::vector<Value> args) {
    if (args.empty()) {
        throw std::runtime_error("syscall requires at least 1 argument");
    }

    long num = args[0].as_int();
    long ret;

    auto arg = [&](int n) -> long long {
        if (auto s = std::get_if<std::string>(&args[n].data)) {
            return reinterpret_cast<long long>(s->c_str());
        } else {
            return std::get<long long>(args[n].data);
        }
    };

    switch (args.size()) {
        case 1:
            ret = syscall(num);
            break;
        case 2:
            ret = syscall(num, arg(1));
            break;
        case 3:
            ret = syscall(num, arg(1), arg(2));
            break;
        case 4:
            ret = syscall(num, arg(1), arg(2), arg(3));
            break;
        case 5:
            ret = syscall(num, arg(1), arg(2), arg(3), arg(4));
            break;
        case 6:
            ret = syscall(num, arg(1), arg(2), arg(3), arg(4), arg(5));
            break;
        case 7:
            ret = syscall(num, arg(1), arg(2), arg(3), arg(4), arg(5), arg(6));
            break;
        default:
            throw std::runtime_error("syscall takes at most 7 arguments");
    }

    return (long long)ret;
}

static Value c_strlen(std::vector<Value> args) {
    if (args[0].type() == Type::STRING) {
        return (long long)std::get<std::string>(args[0].data).size();
    } else {
        return 0;
    }
}

static Value c_typeof(std::vector<Value> args) {
    switch (args[0].type()) {
        case Type::BOOL: {
            return "bool";
        }
        case Type::FLOAT: {
            return "float";
        }
        case Type::STRING: {
            return "string";
        }
        case Type::INT: {
            return "int";
        }
    }
}

static Value c_platformid(std::vector<Value> args) {
#if defined(__APPLE__)
    return "darwin";
#elif defined(__linux__)
    return "linux";
#elif defined(_WIN32) || defined(_WIN64)
    return "windows";
#endif
}

static Value c_sin(std::vector<Value> args) { return std::sin(args[0].as_float()); }
static Value c_cos(std::vector<Value> args) { return std::cos(args[0].as_float()); }
static Value c_tan(std::vector<Value> args) { return std::tan(args[0].as_float()); }
static Value c_abs(std::vector<Value> args) { return std::abs(args[0].as_float()); }
static Value c_sqrt(std::vector<Value> args) { return std::sqrt(args[0].as_float()); }
static Value c_log(std::vector<Value> args) { return std::log(args[0].as_float()); }
static Value c_log10(std::vector<Value> args) { return std::log10(args[0].as_float()); }

Environment* build_env() {
    auto env = new Environment();
    env->vars = {};
    env->ufnlut = {};

    env->bfnlut = {
        { "println", { -1, c_println } },
        { "print", { -1, c_print } },
        { "input", { 1, c_input } },
        { "syscall", { -1, c_syscall } },
        { "strlen", { 1, c_strlen } },
        { "typeof", { 1, c_typeof } },
        { "platformid", { 0, c_platformid } },
        { "sin", { 1, c_sin } },
        { "cos", { 1, c_cos } },
        { "tan", { 1, c_tan } },
        { "abs", { 1, c_abs } },
        { "sqrt", { 1, c_sqrt } },
        { "log", { 1, c_log } },
        { "log10", { 1, c_log10 } }
    };

    auto pbtns = [&](std::vector<Value> args) -> Value {
        std::cout << "name | nargs\n";
        for (const auto& bfn : env->bfnlut) {
            std::cout << bfn.first << " | " << bfn.second.nargs << "\n";
        }
        return 0;
    };

    env->bfnlut.insert({ "print_builtins", { 0, pbtns } });

    return env;
}

File: src/build_env.rs
use crate::eval::{BuiltinFn, Environment};
use crate::value::Value;
use libc::syscall;
use usr_input::input;

fn c_print(args: Vec<Value>) -> Value {
    for arg in args {
        print!("{}", arg);
    }
    Value::Int(0)
}

fn c_println(args: Vec<Value>) -> Value {
    c_print(args);
    println!();
    Value::Int(0)
}

fn c_input(args: Vec<Value>) -> Value {
    let mut buf = String::new();
    input!(buf, "{}", args[0]).expect("Failed to get input");
    buf.truncate(buf.trim_end().len());
    Value::String(buf)
}

fn c_strlen(args: Vec<Value>) -> Value {
    if let Value::String(str) = &args[0] {
        Value::Int(str.len() as i64)
    } else {
        Value::Int(0)
    }
}

fn c_typeof(args: Vec<Value>) -> Value {
    match args[0] {
        Value::Int(_) => Value::String("int".to_string()),
        Value::Float(_) => Value::String("float".to_string()),
        Value::String(_) => Value::String("string".to_string()),
        Value::Bool(_) => Value::String("bool".to_string()),
        Value::Array(_) => Value::String("array".to_string()),
    }
}

fn c_platformid(_: Vec<Value>) -> Value {
    Value::String(std::env::consts::OS.to_string())
}

fn c_syscall(args: Vec<Value>) -> Value {
    let num = args[0].as_int() as i32;
    let arg = |n: usize| args[n].as_syscall_arg();

    let ret = unsafe {
        match args.len() {
            1 => syscall(num),
            2 => syscall(num, arg(1)),
            3 => syscall(num, arg(1), arg(2)),
            4 => syscall(num, arg(1), arg(2), arg(3)),
            5 => syscall(num, arg(1), arg(2), arg(3), arg(4)),
            6 => syscall(num, arg(1), arg(2), arg(3), arg(4), arg(5)),
            7 => syscall(num, arg(1), arg(2), arg(3), arg(4), arg(5), arg(6)),
            _ => panic!("syscall takes at most 7 arguments"),
        }
    };

    Value::Int(ret as i64)
}

fn c_sin(args: Vec<Value>) -> Value {
    Value::Float(args[0].as_float().sin())
}

fn c_cos(args: Vec<Value>) -> Value {
    Value::Float(args[0].as_float().cos())
}

fn c_tan(args: Vec<Value>) -> Value {
    Value::Float(args[0].as_float().tan())
}

fn c_abs(args: Vec<Value>) -> Value {
    Value::Float(args[0].as_float().abs())
}

fn c_sqrt(args: Vec<Value>) -> Value {
    Value::Float(args[0].as_float().sqrt())
}

fn c_log(args: Vec<Value>) -> Value {
    Value::Float(args[0].as_float().log2())
}

fn c_log10(args: Vec<Value>) -> Value {
    Value::Float(args[0].as_float().log10())
}

macro_rules! mkbpair {
    ($name:literal, $nargs:expr, $fn:ident) => {
        (
            $name.to_string(),
            BuiltinFn {
                nargs: $nargs,
                fnc: $fn,
            },
        )
    };
}

pub fn build_env() -> Option<Environment> {
    let mut env = Environment::default();

    env.bfnlut.extend([
        mkbpair!("print", -1, c_print),
        mkbpair!("println", -1, c_println),
        mkbpair!("input", 1, c_input),
        mkbpair!("syscall", -1, c_syscall),
        mkbpair!("strlen", 1, c_strlen),
        mkbpair!("typeof", 1, c_typeof),
        mkbpair!("platformid", 0, c_platformid),
        mkbpair!("sin", 1, c_sin),
        mkbpair!("cos", 1, c_cos),
        mkbpair!("tan", 1, c_tan),
        mkbpair!("abs", 1, c_abs),
        mkbpair!("sqrt", 1, c_sqrt),
        mkbpair!("log", 1, c_log),
        mkbpair!("log10", 1, c_log10),
    ]);

    Some(env)
}
